### PR TITLE
[TIKA-3418] DefaultZipContainerDetector does not support loading of ZipContainerDetectors in an OSGi enviroment

### PR DESCRIPTION
**Pull request by simon-shlomo**
Thanks for your contribution to [Apache Tika](https://tika.apache.org/)! Your help is appreciated!

Before opening the pull request, please verify that
* there is an open issue on the [Tika issue tracker](https://issues.apache.org/jira/projects/TIKA) which describes the problem or the improvement. We cannot accept pull requests without an issue because the change wouldn't be listed in the release notes.
* the issue ID (`TIKA-XXXX`)
  - is referenced in the title of the pull request
  - and placed in front of your commit messages surrounded by square brackets (`[TIKA-XXXX] Issue or pull request title`)
* commits are squashed into a single one (or few commits for larger changes)
* Tika is successfully built and unit tests pass by running `mvn clean test`
* there should be no conflicts when merging the pull request branch into the *recent* `main` branch. If there are conflicts, please try to rebase the pull request branch on top of a freshly pulled `main` branch.

We will be able to faster integrate your pull request if these conditions are met. If you have any questions how to fix your problem or about using Tika in general, please sign up for the [Tika mailing list](http://tika.apache.org/mail-lists.html). Thanks!

### CODE REVIEW

**Comment by simon-shlomo** [comment]
Hi
Could someone please approve the "workflow"?
Have a great weekend
Simon

File: tika-core/src/main/java/org/apache/tika/config/ServiceLoader.java [hunk 1]
```diff
@@ -76,6 +76,11 @@ public ServiceLoader(ClassLoader loader, LoadErrorHandler handler) {
         this(loader, handler, false);
     }
 
+    public ServiceLoader(ClassLoader loader, boolean dynamic) {
+        this(loader, Boolean.getBoolean("org.apache.tika.service.error.warn") ? LoadErrorHandler.WARN :
+                        LoadErrorHandler.IGNORE, true);
```
Comment (tballison) [inline tika-core/src/main/java/org/apache/tika/config/ServiceLoader.java:81]: Should `true` be changed to `dynamic`?  Do we need this constructor?

File: tika-core/src/main/java/org/apache/tika/config/TikaActivator.java [hunk 2]
```diff
@@ -51,14 +53,17 @@ public void start(final BundleContext context) throws Exception {
 
         detectorTracker = new ServiceTracker(context, Detector.class.getName(), this);
         parserTracker = new ServiceTracker(context, Parser.class.getName(), this);
+        zipDetectorTracker = new ServiceTracker(context, "org.apache.tika.detect.zip.ZipContainerDetector", this);
```
Comment (tballison) [inline tika-core/src/main/java/org/apache/tika/config/TikaActivator.java:56]: Do we need to do this for EncodingDetector as well to ensure that all encoding detectors are dynamically loaded?  Obv, separate pull request...but I'm curious if that is failing in OSGi.

File: tika-parsers/tika-parsers-standard/tika-parsers-standard-modules/tika-parser-zip-commons/src/main/java/org/apache/tika/detect/zip/DefaultZipContainerDetector.java [hunk 3]
```diff
@@ -262,12 +266,21 @@ private MediaType detect(ZipArchiveEntry zae, ZipArchiveInputStream zis,
...
             if (mt != null) {
                 return mt;
             }
         }
         return MediaType.APPLICATION_ZIP;
     }
+    
+    private List<ZipContainerDetector> getDetectors() {
+        if (loader != null) {
+            List<ZipContainerDetector> zipDetectors = new ArrayList<>(staticZipDetectors);
```
Comment (tballison) [inline tika-parsers/tika-parsers-standard/tika-parsers-standard-modules/tika-parser-zip-commons/src/main/java/org/apache/tika/detect/zip/DefaultZipContainerDetector.java:280]: In DefaultDetector, the order is dynamic detectors and then we add the statically loaded detectors with `super.getDetectors()`.  Here the order is reversed.  Why the difference, will it make any difference?

**Review by tballison — COMMENTED** [review]
I'm sorry for my delay in reviewing this.  Thank you for identifying this problem and offering a solution!

**Comment by tballison** [comment]
I merged this and made a few modifications.  I need to figure out how to do a better job of passing tikaconfig info into DefaultZipDetector.

Thank you @simon-shlomo for opening this!

**Pull request closed without merging**